torch/csrc/jit/passes/utils/subgraph_matcher.h

```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+#include <torch/csrc/jit/ir.h>
+
+#include <iostream>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace torch {
+namespace jit {
+
+// Usage: instantiate a Graph and use a MatchIterator to iteratively
+// get matches on the underlying graph.  The returned `Match`s can
+// be used as maps from the passed in match graph to the graph the
+// match was run on.
+// 
+//   for (auto m : MatchIterator(match_graph, graph)) {
+//     AT_ASSERT(m[m0.value()] == a.value());
+//   }
+//
+// Optionally, pass in a std::ostream* to get debug information.
+// e.g. `MatchIterator(match_graph, graph, &std::cout)`.
+//
```

Comment:
I think the style for class/struct comments we're planning to eventually converge to is:
```
/** \brief Short description.
  *
  * Verbose description.
  */
```